### PR TITLE
WIP: #156 Preview of thread support.

### PR DESCRIPTION
**Pull request by RobertZenz**
This is just a preview of the upcoming thread support and is not done yet.

The goal is to provide support to use an `Expression` in multiple threads with different variables/functions/operators. For this a new class `Scope` is introduced, which holds the variables/functions/operators of each `Expression`.

### CODE REVIEW

**Comment by RobertZenz** [comment]
Feedback is welcome.

I plan on adding some sort of "chaining" to scopes and I want to extract a "DefaultScope" which contains all the default operators and functions.

**Comment by RobertZenz** [comment]
I believe this is as done as it gets. I mean, the basics are done. Please review this when you find the time and give me feedback on it, whether it is something you want (in one way or another), or not. I'm not sure about it, to be honest, there are a few interesting ideas in here, but whether this implementation is something that should stay, I can not say.

First, let me start with the positive things. Making `Expression` thread-safe was easy. The new `Scope` class is just a stupid container for functions, variables and operators. `Expression`s can now be evaluated by passing an additional `Scope` to them providing variables, functions and operators. This allows to use the same `Expression` for different input sets.

There are a few downsides. For example the additional `Scope` does not know anything about the `Expression`, so it cannot access the `MathContext`, for example, which means that you can't add a number as String to a `Scope`, it must be a `BigDecimal` or `LazyNumber`. The logic for retrieving numbers, functions and operators is a little bit more complicated, obviously.

**Comment by RobertZenz** [comment]
I've been thinking about this further the last few days and tried to implement a similar idea which I called "context support". But the truth is, there is nothing to be gained without changing the API quite substantially...and there's nothing to gain with that either.

My main idea was to have a `Context` which allows to gather functions/operators/variables. There would be the default `Context` which contains all the default functions/operators/variables. A constructor would allow to override this with a custom one, like this:

```java
public Expression() {
    this(new DefaultContext());
)

public Expression(Context context) {
    super();

    this.context = context;
}
```

That would allow to control the available functions rather fine-tuned. But that would change the internals quite a bit. Also, it would mean that the `Context` must be changed for every instance of the `Expression`, which is a rather weird way to look at it.

---

Long story short, after a lot of thinking, I'm closing this as it adds *nothing* to the mix in a real-world scenario.

What *might* be interesting to do in the future would be to revise the API as a whole. Splitting everything up into their own classes and implementations and providing a revised API. I might be working on that for fun, but I don't expect it to make it into EvalEx in the long run. From time to time it's just fun to refactor things and see where that leads, like this.
